### DISCUSSION TITLE
Reducing chunk size while using Firebase snapshot subscriber wrapped around app

### DISCUSSION THREAD

**Discussion by awlevin** [vercel/next.js#13886] [Help] [answered]
Hi all. Specific question is at the bottom if you want to skip to it. Including some context below:

We're migrating an existing CRA to NextJS and trying to _really_ optimize our pages along the way since SEO is a huge priority. 

In our app, we have a Firebase snapshot subscription at the root level, subscribed to a permissions document that belongs to the user (our RBAC). Obviously, some parts of the UI depend on this (premium content, private & admin portions of the app, etc.). The Next.JS app even uses this approach [here](https://github.com/vercel/next.js/blob/canary/examples/with-firebase/pages/_app.js). 

The "problem" is that when you build this, all of the chunks/paths end up including the Firebase module in their "First Load JS" and thus end up much larger than necessary. We have a CMS portion of the site where we especially want these page sizes to be as small as possible to optimize our SEO performance.

**My question:** is there a good approach we can use to include a permissions snapshot listener without requiring Firebase wrapping every page of the app? For example, our CMS pages have an "edit" button if a person has permission to edit that page, but for 99% of people viewing the page, their permissions won't be relevant (most people are viewers and not editors, esp. with respect to our users coming from organic search), so we want the page to load & display everything as fast as possible, without the Firebase module. One thing I have in mind would be to lazy load a component with Firebase and an auth provider context at the `_app.js` level, and the fallback would just be the children of the auth context without the provider. Something like `<Suspense fallback={children}><AuthProvider>{children}</AuthProvider></Suspense>`. I know I can try this out and see for myself but we're still in the early stages of this migration so I figured I'd rather see what other people think / what other solutions are out there before I go down any rabbit holes. 

Thanks!

**Comment by awlevin** [comment, marked as answer]
I ended up lazy loading a component in a layout that we wrap around our app in `_app.tsx`, with all of our firebase logic and connected to our Redux store. Now all of our pages are less than half the initial load size. Unfortunately, this approach causes some of the UI to briefly flash while it transitions from "logged out" to "logged in," for example, if the user is logged in but the lazy component hasn't loaded & the Firebase auth listener hasn't updated Redux with the account information yet. For our use case, since this UI that "flashes" is tucked away in the navbar and isn't visible upon initial load, this isn't a huge issue. This is also only an issue on first load— subsequent client side re-renders seem to work perfectly (as expected). 

Here's what it looks like:

```jsx
// _app.tsx
const ReduxAuthWrapper = dynamic( () => import('../../utils/auth/ReduxAuthWrapper'), { ssr: false })

export default ({ children }: any) => {
  const isServer = typeof window === 'undefined';
  if (isServer) {
    return <Layout>{children}</Layout>
  } else {
    return (
      <>
        <ReduxAuthWrapper />
        <Layout>
          {children}
        </Layout>
      </>
    )
  }
}
```

Edit: I should also clarify that the `ReduxAuthWrapper` component doesn't return any children (only a fragment). This feels wrong— so if anyone has a better suggestion for lazy loading (like lazy loading a hook?) I'd be up for it.

**Comment by nandorojo** [comment]
I've been curious about the same thing. What about using firebase throughout an app, do we have to use `next/dynamic` in every location I import `firebase`? I figure I'll isolate my usage into specific hooks, and lazy load those hooks, but not sure if lazy loading is supported beyond components or not.

**Reply by nandorojo** [reply]
Answered my own question here, kind of: https://next-site-git-new-docs.zeit.now.sh/learn/excel/lazy-loading-modules/test-results seems like the package only gets imported once.

**Reply by Marigameo** [reply]
Do you have working example for how to use next/dynamic for loading firebase import dynamically. The link looks like broken. Can you share it here?
